add changeworld listener

456dev opened the pull request:
fixes #15 (probably, i didnt have a foolproof reproduction)

456dev committed:
add changeworld listener

tastybento commented on src/main/java/world/bentobox/parkour/listeners/CourseRunnerListener.java:
```diff
@@ -71,11 +72,40 @@ public void onVisitorArrive(IslandEnterEvent e) {
-        } else {
-            user.setGameMode(GameMode.SURVIVAL);
+
+        if (addon.inWorld(user.getLocation())) {
```
This includes nether and end. So if you are trying to keep it to the same world then you'll need to check some other things too.

456dev replied:
It probably doesn't matter here, since both the to and the from world's are managed by parkour

456dev replied:
clarified comments a little

tastybento commented on src/main/java/world/bentobox/parkour/listeners/CourseRunnerListener.java:
```diff
@@ -71,11 +72,40 @@ public void onVisitorArrive(IslandEnterEvent e) {
+    public void onPlayerChangeWorld(PlayerChangedWorldEvent event) {
+        User user = User.getInstance(event.getPlayer());
+        Location currentLocation = user.getLocation();
+        if (!addon.inWorld(currentLocation)) {
```
This includes nether and end.

456dev replied:
If a player changes world into any parkour managed world, it should still change gamenode, right?

tastybento replied:
Yes. I'm just thinking if you need to consider the use case where someone is in the Parkour Nether, but teleports back to the Parkour Overworld. I anticipate Parkour courses that cross dimensions. That's why.

456dev replied:
At the moment, cross dimensional travel is blocked by the teleport event iirc, is that also a bug?

456dev replied:
stopped portal teleports from stopping run, added a check that stops changing the gamemode if you are coming from a parkour world

tastybento commented on src/main/java/world/bentobox/parkour/listeners/CourseRunnerListener.java:
```diff
@@ -71,11 +72,40 @@ public void onVisitorArrive(IslandEnterEvent e) {
+        }
+        // Check if player is in a parkour course
+        Optional<Island> course = addon.getIslandsManager().getIslandAt(currentLocation);
+        if (course.isEmpty()) {
```
If you use isPresent, then you don't need to check isEmpty or make the local variable "course".

456dev replied:
done

tastybento reviewed (changes requested):
The other way I think you could do this is put the game mode changes into a `Bukkit.getScheduler().runTask` and have them execute a tick afterwards, which will be when they arrive in the next world. But I'm assuming that it will take just a tick (in my experience this is always the case, but...). So your approach is probably cleaner.

456dev committed:
simplify logic, add notification for run exit on death
remove portals from canceling run, ensure gamemode is only changed in certain teleport events

make it more explicit that IslandEnterEvent checks source world before changing gamemode

non-parkour -> parkour: PlayerChangedWorldEvent set gm
non-parkour -> non-parkour: none
parkour -> parkour:
        same island: PlayerTeleportEvent set gm if not running
        between islands: IslandEnterEvent set gm
parkour -> non-parkour: none

tastybento committed:
Merge branch 'develop' into the456gamer-crossworld-creative

tastybento commented:
A few tests fail, but I think they are fixable. I'll merge.
